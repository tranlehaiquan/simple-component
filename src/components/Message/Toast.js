import React, { useEffect } from 'react';
import { CSSTransition } from 'react-transition-group';
import propTypes from 'prop-types';
import classnames from 'classnames';
import noop from 'lodash/noop';

import Icon from '../Icon/Icon';

function Toast(props) {
  const { message, type, show, customClass, close, duration, onClose } = props;
  const toastClassName = classnames('p-toast', type && `p-toast--${type}`, customClass);
  let timer;

  const clearTimer = () => {
    clearTimeout(timer);
  };

  const setTimer = () => {
    if(duration && close) {
      timer = setTimeout(close, duration);
    }
  };

  const handleClose = () => {
    close();
  };

  useEffect(() => {
    setTimer();
    return clearTimer;
  }, []);

  return (
    <CSSTransition 
      timeout={300}
      in={show} 
      unmountOnExit
      classNames="toast"
      onEntered={setTimer}
      onExit={clearTimer}
      onExited={onClose}
    >
      {() => (
        <div className={toastClassName} onMouseEnter={clearTimer} onMouseLeave={setTimer} >
          <p className="p-toast__description">
            {type &&
              <Icon name={typeIcons[type]} className={`p-toast__icon p-toast__icon--${type}`}/>
            }
            { message }
          </p>
          {close &&
            <span className="p-toast__close" onClick={handleClose}>
              <Icon name="CLOSE" />
            </span>
          }
        </div>
      )}
    </CSSTransition>
  );
}

const typeIcons = {
  'success': 'CHECKED',
  'warning': 'INFO',
  'info': 'INFO',
  'danger': 'CANCEL'
};

Toast.propTypes = {
  type: propTypes.oneOf(['success', 'info', 'warning', 'danger']),
  message: propTypes.string,
  show: propTypes.bool,
  customClass: propTypes.string,
  close: propTypes.func,
  duration: function(props, propName, componentName) {
    if(!props[propName]) return;
    const typeOfProp = typeof props[propName];
    if(typeOfProp !== 'number') {
      return new Error(
        `Failed prop type: Invalid prop duration of type ${typeOfProp} supplied to ${componentName}, expected number.`);
    }

    if(!props.close) {
      // eslint-disable-next-line
      console.warn(`Please supplie close method with duration prop to ${componentName}`);
    }
  },
  onClose: propTypes.func
};

Toast.defaultProps = {
  type: 'info',
  message: '',
  show: false,
  customClass: '',
  onClose: noop
};

export default Toast;
