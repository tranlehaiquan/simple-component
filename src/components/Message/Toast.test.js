import React from 'react';
import renderer from 'react-test-renderer';
import Toast from './Toast';
import { shallow, mount } from 'enzyme';

test('render snapshot Toast', () => {
  const component = renderer.create(
    <Toast 
      message="Hello the world???" 
      show={true} 
      type="success"
      close={() => {}}
      duration={1500}
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
