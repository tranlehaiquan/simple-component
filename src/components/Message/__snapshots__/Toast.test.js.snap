// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render snapshot Toast 1`] = `
<div
  className="p-toast p-toast--success"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <p
    className="p-toast__description"
  >
    <svg
      className="sp-icon p-toast__icon p-toast__icon--success"
      height={22}
      viewBox="0 0 1024 1024"
      width={22}
    >
      <path
        d="M426 726l384-384-60-62-324 324-152-152-60 60zM512 86c236 0 426 190 426 426s-190 426-426 426-426-190-426-426 190-426 426-426z"
      />
    </svg>
    Hello the world???
  </p>
  <span
    className="p-toast__close"
    onClick={[Function]}
  >
    <svg
      className="sp-icon"
      height={22}
      viewBox="0 0 1024 1024"
      width={22}
    >
      <path
        d="M810 274l-238 238 238 238-60 60-238-238-238 238-60-60 238-238-238-238 60-60 238 238 238-238z"
      />
    </svg>
  </span>
</div>
`;
